common/expected.hpp

```diff
@@ -0,0 +1,123 @@
+// Copyright 2023 Northern.tech AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#include <cassert>
+#include <string>
+#include <type_traits>
+
+#ifndef EXPECTED_HPP
```

Comment:
Related to https://github.com/mendersoftware/mender/pull/1115#discussion_r1063173922, we should namespace this. I used `mender_common` in my example, mostly to avoid the extremely generic `common`. It depends a bit on the outcome of that discussion though. If you think it's worthwhile to export our own `std::expected` variant if it's not available, then it doesn't really matter if the namespace is long and ugly, because we won't use it directly.